clang/lib/Sema/BoundsWideningAnalysis.cpp

```diff
@@ -14,4 +14,1213 @@
+  // Add variables occurring in StmtGen for the current statement to the list
+  // of variables that are pointers to null-terminated arrays.
+  for (auto VarBoundsPair : EB->StmtGen[CurrStmt])
+    AllNtPtrsInFunc.insert(VarBoundsPair.first);
+}
+
+void BoundsWideningAnalysis::FillStmtGenKillSets(ElevatedCFGBlock *EB,
+                                                 const Stmt *CurrStmt,
+                                                 BoundsMapTy &VarsAndBounds) {
+  for (auto VarBoundsPair : VarsAndBounds) {
+    const VarDecl *V = VarBoundsPair.first;
+    RangeBoundsExpr *R = VarBoundsPair.second;
+
+    EB->StmtGen[CurrStmt][V] = R;
+    EB->StmtKill[CurrStmt].insert(V);
+  }
+}
+
+void BoundsWideningAnalysis::GetVarsAndBoundsInDecl(
+  ElevatedCFGBlock *EB, const VarDecl *V, BoundsMapTy &VarsAndBounds) {
+
+  if (BWUtil.IsNtArrayType(V)) {
+    BoundsExpr *NormalizedBounds = SemaRef.NormalizeBounds(V);
+    VarsAndBounds[V] = dyn_cast<RangeBoundsExpr>(NormalizedBounds);
+  }
+}
+
+void BoundsWideningAnalysis::GetVarsAndBoundsInWhereClause(
+  ElevatedCFGBlock *EB, WhereClause *WC, BoundsMapTy &VarsAndBounds) {
+
+  if (!WC)
+    return;
+
+  for (const auto *Fact : WC->getFacts()) {
+    auto *BF = dyn_cast<BoundsDeclFact>(Fact);
+    if (!BF)
+      continue;
+
+    VarDecl *V = BF->Var;
+    if (BWUtil.IsNtArrayType(V)) {
+      BoundsExpr *NormalizedBounds = SemaRef.ExpandBoundsToRange(V, BF->Bounds);
+      VarsAndBounds[V] = dyn_cast<RangeBoundsExpr>(NormalizedBounds);
+    }
+  }
+}
+
+void BoundsWideningAnalysis::GetVarsAndBoundsInPtrDeref(
+  ElevatedCFGBlock *EB, BoundsMapTy &VarsAndBounds) {
+
+  // Get the terminating condition of the block.
+  // The CFG for a compound condition like "if (e1 && e2 && e3)" is as follows:
+
+  // B1->B2 (edge condition e1)
+  // B2->B3 (edge condition e2)
+  // B3->B4 (edge condition e3)
+
+  // So we see that each condition becomes an edge in the CFG.
+  // getTerminatorStmt() would give us the entire condition "e1 && e2 && e3"
+  // while getLastCondition() would give us e1 for the first edge e2 for the
+  // second and so on.
+  const Expr *TermCond = EB->Block->getLastCondition();
+  if (!TermCond)
+    return;
+
+  // If the terminating condition is a dereference expression get that
+  // expression. For example, from a terminating condition like "if (*(p +
+  // 1))", extract the expression "p + 1".
+  Expr *DerefExpr = BWUtil.GetDerefExpr(TermCond);
+  if (!DerefExpr)
+    return;
+
+  // Store the dereference condition for the block. We will use it later in the
+  // In set calculation.
+  EB->TermCondDerefExpr = DerefExpr;
+
+  // Get all variables in the expression that are pointers to null-terminated
+  // arrays. For example: On a dereference expression like "*(p + i + j + 1)"
+  // GetNtPtrsInExpr() will return {p} if p is a pointer to a null-terminated
+  // array.
+
+  VarSetTy NtPtrsInExpr;
+  BWUtil.GetNtPtrsInExpr(DerefExpr, NtPtrsInExpr);
```

Comment:
This function should return a single NtPtr because a deref expr can only contain at most one null-terminated pointer.